docs/source/torch.compiler_faq.rst

```diff
@@ -528,6 +528,162 @@ invokes an ``nn.Module``. This is because the outputs now depend on the
+-------------------------------------------------------
+
+Yes you can! To do so, you may simply execute your code under ``torch.device("cuda")``
+
+.. code-block:: python
+
+   @torch.compile
+   def numpy_fn(X: ndarray, Y: ndarray): -> ndarray
+       # Compute the ndarray Z here
+       return Z
+
+
+   X = np.random.randn(1000, 1000)
+   Y = np.random.randn(1000, 1000)
+   with torch.device("cuda"):
+       Z = numpy_fn(X, Y)
+
+In this example, ``numpy_fn`` will be executed in CUDA. For this to be
+possible, ``torch.compile`` automatically moves ``X`` and ``Y`` from CPU
+to CUDA, and then it moves the result ``Z`` from CUDA to CPU. If we are
+executing this function several times in the same program run, we may want
+to avoid all these rather expensive memory copies. To do so, we just need
+to tweak our ``numpy_fn`` so that it accepts cuda Tensors and returns tensors:
+
+.. code-block:: python
+
+   @torch.compile
+   def numpy_fn(X: Tensor, Y: Tensor): -> Tensor
+       X = X.numpy()
+       Y = Y.numpy()
+       # Compute Z here
+       Z = torch.from_numpy(Z)
+       return Z
+
+   X = torch.randn(1000, 1000, device="cuda")
+   Y = torch.randn(1000, 1000, device="cuda")
+   with torch.device("cuda"):
+       Z = numpy_fn(X, Y)
+
+By doing this, we explicitly create the tensors in CUDA memory, and we keep
+them there. In this case ``X.numpy()`` and ``from_numpy()`` are hints to the compiler
+but no real data movement happens. Note that the original program would not run
+on eager mode now. If you want to run it in eager mode, you would need to call
+``.numpy(force=True)`` doing ``Z = Z.cuda()`` before returning
+``Z``. Of course, doing this would execute the program on eager mode NumPy, and
+on CPU.
+
+
+How do I debug my ``torch.compile``d NumPy code?
+------------------------------------------------
+
+Debugging JIT compiled code is challenging, given the complexity of modern
+compilers and the daunting errors that they raise.
+`The tutorial on how to diagnose runtime errors within torch.compile <https://pytorch.org/docs/main/torch.compiler_troubleshooting.html#diagnosing-runtime-errors>`__
+contains a few tips and tricks on how to tackle this task.
+
+If the above is not enough to pinpoint the origin of the issue, there are still
+a few other NumPy specific tools we can use. If we find a bug that is blocking
+our development and we are tracing through mixed PyTorch <> NumPy code where
+the NumPy part is not particularly bulky, we can simply deactivate the NumPy
+tracing altogether by doing
+
+.. code-block:: python
+
+   from torch._dynamo import config config.trace_numpy = False
+
+This moves back to the behavior in 2.0 and will avoid tracing through any NumPy
+function.
+
+Of course, this is not a satisfactory answer if our program is mostly composed
+of NumPy code. In these cases, we can try to execute eagerly (without
+``torch.compile``) the NumPy code on PyTorch by importing ``import torch._numpy as np``.
```

Comment:
Just a thought, but this may be a useful tool in and of itself. If you could enable a flag and dynamo will not inline from `torch._numpy` but will do the conversion from numpy to `torch._numpy`. That way you don't have to edit any code at all to do this test.

Reply:
The point of changing the imports is that, in most cases, you can execute the code without using dynamo at all. Then, if you see the same bug while doing this, the bug is certainly in the decomposition.

I found this *very* helpful when debugging issues, as when doing this you are dealing with very simple Python code and debugging stuff is as easy as it gets.